### PR TITLE
LUCENE-10603: Migrate remaining SSDV iteration to use docValueCount in production code

### PR DESCRIPTION
**Pull request by gsmiller**
This migrates the remaining production code iteration to use `SSDV#docValueCount` for iteration, getting us closer to removing support for `NO_MORE_ORDS` in `SSDV#nextOrd`. Test-related code still needs to get updated.

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/index/CheckIndex.java [hunk 1]
```diff
@@ -3382,6 +3383,7 @@ private static void checkSortedSetDocValues(
         seenOrds.set(ord);
         ordCount++;
       }
+
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/index/CheckIndex.java:3386]: At this point `ord` is going to be the last ord while it used to always be NO_MORE_ORDS, which I suspect may cause CheckIndex failures.

File: lucene/core/src/java/org/apache/lucene/search/SortedSetSelector.java [hunk 2]
```diff
@@ -226,12 +226,14 @@ public int lookupTerm(BytesRef key) throws IOException {
 
     private void setOrd() throws IOException {
       if (docID() != NO_MORE_DOCS) {
-        while (true) {
-          long nextOrd = in.nextOrd();
-          if (nextOrd == NO_MORE_ORDS) {
-            break;
+        int docValueCount = in.docValueCount();
+        if (docValueCount == 0) {
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/search/SortedSetSelector.java:230]: docValueCount may never return 0, we can drop this branch

File: lucene/core/src/java/org/apache/lucene/search/SortedSetSelector.java [hunk 3]
```diff
@@ -304,25 +306,19 @@ public int lookupTerm(BytesRef key) throws IOException {
...
-          }
-          if (upto == ords.length) {
-            ords = ArrayUtil.grow(ords);
-          }
-          ords[upto++] = (int) nextOrd;
-        }
-
-        if (upto == 0) {
+        int docValueCount = in.docValueCount();
+        if (docValueCount == 0) {
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/search/SortedSetSelector.java:310]: likewise here

File: lucene/core/src/java/org/apache/lucene/search/SortedSetSelector.java [hunk 4]
```diff
@@ -304,25 +306,19 @@ public int lookupTerm(BytesRef key) throws IOException {
...
-        } else {
-          ord = ords[(upto - 1) >>> 1];
+          return;
+        }
+
+        ords = ArrayUtil.grow(ords, docValueCount);
+        for (int i = 0; i < docValueCount; i++) {
+          ords[i] = (int) in.nextOrd();
         }
+        ord = ords[(docValueCount - 1) >>> 1];
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/search/SortedSetSelector.java:321]: we don't even need to buffer ords now that we know their number I think?
We could compute the index of the ord we're interested in, then consume this number of ords, and the next ord would be the median ord?

File: lucene/core/src/java/org/apache/lucene/search/SortedSetSelector.java [hunk 5]
```diff
@@ -394,25 +390,19 @@ public int lookupTerm(BytesRef key) throws IOException {
...
-        } else {
-          ord = ords[upto >>> 1];
+          return;
+        }
+
+        ords = ArrayUtil.grow(ords, docValueCount);
+        for (int i = 0; i < docValueCount; i++) {
+          ords[i] = (int) in.nextOrd();
         }
+        ord = ords[docValueCount >>> 1];
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/search/SortedSetSelector.java:405]: and likewise here?

File: lucene/core/src/java/org/apache/lucene/search/SortedSetSelector.java [hunk 6]
```diff
@@ -394,25 +390,19 @@ public int lookupTerm(BytesRef key) throws IOException {
...
-          }
-          if (upto == ords.length) {
-            ords = ArrayUtil.grow(ords);
-          }
-          ords[upto++] = (int) nextOrd;
-        }
-
-        if (upto == 0) {
+        int docValueCount = in.docValueCount();
+        if (docValueCount == 0) {
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/search/SortedSetSelector.java:394]: we can drop this branch

**Review by jpountz — COMMENTED** [review]
Great cleanup!

File: lucene/core/src/java/org/apache/lucene/index/CheckIndex.java (hunk 1 above)
Comment (gsmiller) [reply lucene/core/src/java/org/apache/lucene/index/CheckIndex.java]: Ack. Yeah that's the issue. I don't think the equality check between `ord` and `ord2` after this loop makes sense anymore given that there's no guarantee about what value `ord` will be if calling `nextOrd()` more times than advertised by `docValueCount()`, so I removed the check.

File: lucene/core/src/java/org/apache/lucene/search/SortedSetSelector.java (hunk 2 above)
Comment (gsmiller) [reply lucene/core/src/java/org/apache/lucene/search/SortedSetSelector.java]: Ah right. Thanks! Addressed this in all four places.

File: lucene/core/src/java/org/apache/lucene/search/SortedSetSelector.java (hunk 4 above)
Comment (gsmiller) [reply lucene/core/src/java/org/apache/lucene/search/SortedSetSelector.java]: Good point. Tweaked this (and the other location).

**Comment by gsmiller** [comment]
Merging as I've addressed the outstanding feedback and the change is otherwise straight-forward. Thanks @jpountz for the suggestions and review!

**Pull request merged**